src/java.base/share/classes/jdk/internal/vm/annotation/Stable.java

```diff
@@ -28,56 +28,191 @@
+ * <em>also</em> treated as independent stable variables.  Such a
+ * stable array may contain any type, reference or primitive.
+ * <p>
+ * More generally, if a stable field is declared as an array type with
+ * <em>D</em> dimensions, then all the non-null components of the
+ * array, and of any sub-arrays up to a nesting depth less than
+ * <em>D</em>, are treated as stable variables.  Thus, a stable field
+ * declared as an array potentially defines a tree (of fixed depth
+ * <em>D</em>) containing many stable variables, with each such stable
+ * variable is independently considered for optimization.  In this
+ * way, and depending on program execution, a single {@code Stable}
+ * annotation can potentially create many independent stable
+ * variables.  Since the top-level array reference is always stable,
+ * it is in general a bad idea to resize the array, even while keeping
+ * all existing components unchanged.  (This could be relaxed in the
+ * future, to allow expansion of stable arrays, if there were a use
+ * case that could deal correctly with races.)
+ * <p>
+ * As very simple example, a boolean variable is constant-folded only
+ * when it is set to {@code true}.  Even this simple behavior is
+ * sometimes useful for recording a permanent one-shot state change,
+ * in such a way that the compiler can remove dead code associated
+ * with the initial state.  It is in general a bad idea to reset
+ * such a variable to {@code false}, since compiled code might have
+ * "frozen" the {@code true} value, and will never detect the reset
+ * value.
  * <p>
  * Fields which are declared {@code final} may also be annotated as stable.
  * Since final fields already behave as stable values, such an annotation
  * conveys no additional information regarding change of the field's value, but
- * still conveys information regarding change of additional components values if
+ * still conveys information regarding change of additional component values if
  * the type of the field is an array type (as described above).
  * <p>
- * The HotSpot VM relies on this annotation to promote a non-null (resp.,
- * non-zero) component value to a constant, thereby enabling superior
- * optimizations of code depending on such a value (such as constant folding).
+ * There are special times early in application startup when constant
+ * folding of stable variables is disabled.  Specifically, the VM does
+ * not observe the effects of any {@code @Stable} annotation until
+ * after an AOT cache assembly phase (if any) has run to completion.
+ * Therefore, during these early times, a stable variable can be
+ * changed in any way, just like a regular mutable variable (field or
+ * array component).  If a field is annotated {@code @Stable} but is
+ * treated in this way during AOT cache assembly, that fact must be
+ * clearly stated as a warning on the field declaration.  If there is
+ * no such warning, maintainers can ignore this edge case.
```

Comment:
The warning on the field declaration is a comment?  As a follow on to this, should we define an additional annotation to carry that warning ie `@AssemblyTimeChangable`?  (better naming to be bike shed)